### ISSUE TITLE
[Refactoring] Structured Saveable Scene States

### ISSUE THREAD

**Issue by aimozg** [Kitteh6660/Corruption-of-Champions-Mod#752]
Draft. Edits and comments appreciated.

## Dark past (and present)
Flags suck.

1. Game state is managed in its raw, saveable form.
2. They are `int`-indexed in a single global array.
3. There's no consistent method of saving game/story-related information. Flags are most popular, but there're also status effects, perks, and key items used for that purpose.

## Bright future

A need to know about existence of savefile content arises only when a scene is designed and new objects are to be introduced to the game state.

In all other places, myriads of happy helper methods with easily understandable names dance around ~naked and lusty~, eager to help.

## Design

TBD.

### API requirements

* Save data structure should not depend on field names, class names, or packages where they are declared.

### 1. Refactor scene-related flag usages to methods.

Pick a victim in `kFLAGS`.
Find `flag[]` usages of it.
Create a wrapper methods in a related class.
Replace.

(Optionally) Copy the magic flag declaration as a `private static const` to that class, and _comment_ it out in the `kFLAGS`.

### 2. Design state structure and define it.
Guidelines:
1. Keep it `private`. Low-level access should go through methods.
2. Avoid creating multiple `Boolean` flags for mutually exclusive properties. Join them into multi-state "enum".
3. Keep enums `int`, not `String`. Define a constants, if needed.

### 3. Implement it
1. Add a loadState method that will handle loading from flags or previous versions
2. Add a saveState method that will handle saving both to save object and to flags.

**Comment by brrritssocold** [comment]
Have you looked at the [ISerializable interface](https://github.com/Kitteh6660/Corruption-of-Champions-Mod/blob/master/classes/classes/internals/ISerializable.as)? (See PR #563 for more info)

I created that Interface to allow classes to store / retrieve their own state, classes can also be written so (de)serialization is recursive.

It is currently only used in [VaginaClass.as](https://github.com/Kitteh6660/Corruption-of-Champions-Mod/blob/master/classes/classes/VaginaClass.as#L146) (because i haven't gotten round to rewriting other classes yet)

**Comment by aimozg** [comment]
@brrritssocold Yes. I have a merged version of your interface and a similar API already done by me in Xianxia mod is used in https://github.com/aimozg/Corruption-of-Champions-Mod/tree/revamp-serialization branch. (Added methods to ease savedata upgrading/downgrading)
